### ISSUE TITLE
vmnet helper

### ISSUE THREAD

**Issue by rb6502** [ksherlock/ample#23]
Hi Kelvin,

I'd like to borrow vmnet_helper (and the MAME-side interface glue) for stock MAME.  As you probably noticed, pcap on macOS pretty much only works with wired Ethernet connections and so I've been unable to use or work on any network emulation since I switched daily driving to my MacBook Pro.  I can go the QEMU route and just require running MAME as root for networking, but having an external helper running as root is a much cleaner idea.

**Comment by ksherlock** [comment]
That's fine.  The helper code is [here](https://github.com/ksherlock/ample/blob/master/vmnet_helper/vmnet_helper.c) and the MAME part is [here](https://github.com/ksherlock/mame/tree/ample-258-u2/src/osd/modules/netdev) and [here](https://github.com/ksherlock/mame/commit/3018fbbec133450902c69b289180b46e9eca5970) for the lua build part.  I didn't think MAME would be interested in it since it's a bit hacky.  vmnet has strongly held opinions about the mac address so it has to rewrite ARP and DHCP messages.

**Comment by rb6502** [comment]
Ok.  I'm trying your plain vmnet where I *do* run MAME as root before getting into any potential weirdness with vmnet_helper and it's sending packets OK as far as I can see but nothing is ever received.  (I'm running an emulated Mac with the Fetch FTP client trying to connect to a server on my LAN).  Any ideas on that?

**Comment by ksherlock** [comment]
Are you using a static IP address?  vmnet requires DHCP.  Most packets are blocked until DHCP negotiation.  I can't recall if I ever tested the plain vmnet so it's also possible there's a bug in there. When vmnet starts up, it creates a bridge100 interface so you can use wireshark or tcpdump to monitor that (inbound packets have checksum errors but that's fine).  If that's ok, FTP would probably have to be in passive mode it isn't already.

**Comment by rb6502** [comment]
Yeah, I was using static IP.  I changed to DHCP and it requests an address and gets a response but it doesn't acknowledge it, it just keeps requesting an address.  Maybe Mac OS does care about the checksum?  What's an easy way to test on the A2?

**Comment by ksherlock** [comment]
I haven't tested with MacOS networking.  I'll try to set that up and look into it.  The IP checksums are fixed at some point after wireshark sees them so they're correct by the time they make it into MAME.

For the Apple II, The [ip65 disk](https://github.com/cc65/ip65/releases/tag/2023-06-11) will work with `-sl3 uthernet`

**Comment by rb6502** [comment]
Hmm.  The Apple II when I try it always forces the interface to -1 even if
I fix it in apple2ee.cfg and so nothing works.

**Comment by rb6502** [comment]
Never mind, that was user error :-)  I got DATE65.SYSTEM to get the time and date correctly.

**Comment by ksherlock** [comment]
For the macintosh, are you using MacTCP, OpenTransport, or something else?

**Comment by rb6502** [comment]
I'm running maciivx -nbe enetnb with 7.6.1 that's been upgraded with Open Transport 1.3.1 (you can get the 1.3 installer and 1.3.1 upgrade from Macintosh Garden).

Related to this, I've added this section: https://wiki.mamedev.org/index.php/Driver:Mac_68K#How_do_I_get_files_into_the_emulation.3F

Would be cool if Ample had a drop box for files to send to the Mac and could run hdiutil for you and set it as the -cdrom, but that's pie-in-the-sky and it's not super hard to do manually.

**Comment by ksherlock** [comment]
I was able to track down the Open Transport DHCP issues and it was my fault (and checksum related).  [Here](https://github.com/ksherlock/mame/blob/ample-259-u2-wip/src/osd/modules/netdev/vmnet_common.h) is the updated file.

<img width="428" alt="image" src="https://github.com/ksherlock/ample/assets/964904/195a9785-b59c-43b7-b673-7809f153def5">

**Comment by rb6502** [comment]
Works great!  I was able to connect to both the FTP and netatalk servers on my home server and copy files back and forth.

I'll clean this stuff up and submit it sometime after the next release is locked down.

**Comment by rb6502** [comment]
Bonus screenshot: https://rbelmont.mameworld.info/wp-content/uploads/2023/09/0000.png

**Comment by rb6502** [comment]
So Open Transport works great now, but MacTCP uses BootP instead of DHCP.  I assume there's no way to tell it otherwise?  I understand a lot of modern routers won't bother with it either so it's very much a problem people with physical machines have too.

**Comment by rb6502** [comment]
Oh, interesting.  That's cool that vmnet still supports it.  Someone at Apple must understand the use cases for vmnet aren't all VMWare types of stuff.

QEMU filed a radar with Apple in 2021 that you shouldn't need root to use vmnet, but I don't expect anything to happen.  https://openradar.appspot.com/radar?id=5007417364447232  I believe they were working on some kind of server app similar to vmnet-helper but I don't know where they're at with it.

**Comment by ksherlock** [comment]
I suspect it's more of accident that bootp still works.

I found one more checksum bug - https://github.com/ksherlock/mame/commit/2152bec136ab10a302b7cd7d771c0ff72df2bde4

**Comment by rb6502** [comment]
Thanks!  I'm told 'networksetup -setbootp (interface name)' is the more Apple-approved way to enable it, but I haven't tried any methods yet.

**Comment by rb6502** [comment]
Fair warning: MAME 0.259 will lock up when reading 3.5" disks.  There's a hot fix at https://github.com/mamedev/mame/commit/cf510adc9d34b4ad9a065f3d5bfbcf17115ea06f

**Comment by ksherlock** [comment]
Thanks for the heads up.  I've cherry-picked it into a new build.

I think the networksetup is for controlling the settings as a client (in system settings -> network -> ... configure IPv4 ... On my 10.14 Mac, bootp is still an option but on 13.5 it's gone).

[This](https://github.com/lima-vm/socket_vmnet) was pointed out to me the other day which is similar but uses a unix domain socket.  Maybe it's better, maybe it's not.  I haven't tried but I don't understand how it can work without using the mac address vmnet assigns.  It could be that bridged mode (added in 10.15) is more lenient about that.

**Comment by rb6502** [comment]
I chased down bootp to /System/Library/LaunchDaemons/bootps.plist where it's disabled and SIP won't allow you to change it even as root.  Supposedly you can get launchd to override that by putting your own version in /Library/LaunchDaemons and/or ~/Library/LaunchDaemons but I lack the macOS-foo to get it to work.

**Comment by rb6502** [comment]
Any idea if there's anything in vmnet either on the MAME side or the Apple side that would block AppleTalk traffic?  I installed the latest netatalk 2.2.10 on my Linux systems and the nbplkup utility shows that they're advertising a server but MAME doesn't see it even when the host Mac is connected with wired Ethernet (WiFi is known to interfere with AppleTalk).

I can connect to the server directly with the newer AppleShare client where you can enter a DNS name or IP address of a server.  So AFP over TCP works fine, but just trying to see available servers in the Chooser doesn't show anything (AARP/NBP/the other components of AppleTalk).

**Comment by ksherlock** [comment]
Bad news... It looks like the Ethertalk packets generated over open transport use a [logical link control](https://en.wikipedia.org/wiki/Logical_link_control) / [subnetwork access protocol](https://en.wikipedia.org/wiki/Subnetwork_Access_Protocol) (LLC/SNAP) ethernet header that vmnet does not like. (The Use 802.3 checkbox in the TCP/IP control panel enables the SNAP/LLC header for TCP as well and vmnet doesn't like that either).  If we remove the SNAP/LLC header the packets still don't make it into vmnet so I think it must be filtering out everything except ARP and IP.

**Comment by rb6502** [comment]
That sucks, and arguably most for the IIgs emulation, although System 6 and pre-030 Macs that can't run Open Transport also get hurt.  I guess the long-term solution will be to adopt a compatible built-in LocalTalk bridge like Mini vMac and GSPlus/GSPort have.  Vas wants to do it differently though: have the 8530 speak the correct bit-serial encoding and then a separate device class similar to the existing bitb for serial does the rest.  The Mac guys have done some interesting stuff along those lines recently for real machines ( https://68kmla.org/bb/index.php?threads/releasing-airtalk-wireless-plug-and-play-localtalk-dongle.43838/ for example).

**Comment by rb6502** [comment]
Here's the current status of the cleanup/integration, not including that I want to do something about the network module selection: https://github.com/rb6502/mame/commit/99bfb0d76b0239bac7528898a8c0bc4f135adf63

**Comment by ksherlock** [comment]
I might be biased but it looks good to me.  The vmnet_common stuff should probably be in a namespace.

bootp/dhcp translation shouldn't be that hard and would be handy.

**Comment by rb6502** [comment]
Someone on 68KMLA is writing a router that converts between LToUDP (LocalTalk over UDP, which vMac, Basilisk, and GSPort/GSPlus speak) and EtherTalk.  They said they're using VMNet in 'bridged' mode instead of 'shared' for their router and it obviously passes EtherTalk packets.

The post about it is here: https://68kmla.org/bb/index.php?threads/developing-a-portable-user-land-appletalk-stack.34622/page-4#post-475509 and I asked them for clarification about the VMNet mode further down the thread.

**Comment by ksherlock** [comment]
Interesting.  VMNET_BRIDGED_MODE didn't exist when I originally wrote the helper.  I'll try to play with it this weekend.

**Comment by ksherlock** [comment]
For bridged mode, we need to specify the interface that is bridged, eg:

```
        xpc_dictionary_set_uint64(dict, vmnet_operation_mode_key, VMNET_BRIDGED_MODE);
        xpc_dictionary_set_string(dict, vmnet_shared_interface_name_key, "en1");
```

`vmnet_start_interface` return NULL if there's an error, which it was doing, so that should exit immediately since the status callback will never be executed and it will otherwise hang forever.
